allennlp/training/cross_validation.py

```diff
@@ -0,0 +1,317 @@
+            test_data_loader = self.data_loader_builder.construct(dataset=test_dataset)
+
+            model = copy.deepcopy(self.model)
+            subtrainer = self.trainer_builder.construct(
+                serialization_dir=serialization_dir, data_loader=train_data_loader, model=model
+            )
+
+            fold_metrics = subtrainer.train()
+
+            for metric_key, metric_value in training_util.evaluate(
+                model,
+                test_data_loader,
+                subtrainer.cuda_device,
+                batch_weight_key=self.batch_weight_key,
+            ).items():
+                if metric_key in fold_metrics:
+                    fold_metrics[f"test_{metric_key}"] = metric_value
+                else:
+                    fold_metrics[metric_key] = metric_value
+
+            common_util.dump_metrics(
+                os.path.join(subtrainer._serialization_dir, "metrics.json"), fold_metrics, log=True
+            )
+
+            metrics_by_fold.append(fold_metrics)
+
+        metrics = {}
+
+        for metric_key, fold_0_metric_value in metrics_by_fold[0].items():
+            if isinstance(fold_0_metric_value, float):
+                average = Average()
+                for fold_index, fold_metrics in enumerate(metrics_by_fold):
+                    metric_value = fold_metrics[metric_key]
+                    metrics[f"fold{fold_index}_{metric_key}"] = metric_value
+                    average(metric_value)
+                metrics[f"average_{metric_key}"] = average.get_metric()
+            else:
+                for fold_index, fold_metrics in enumerate(metrics_by_fold):
+                    metrics[f"fold{fold_index}_{metric_key}"] = fold_metrics[metric_key]
+
+        if self.retrain:
+            self.trainer.train()
+
+        return metrics
+
+    @overrides
+    def finish(self, metrics: Dict[str, Any]):
+        common_util.dump_metrics(
+            os.path.join(self.serialization_dir, "metrics.json"), metrics, log=True
+        )
+
+    @classmethod
+    @overrides
+    def from_partial_objects(
+        cls,
+        serialization_dir: str,
+        batch_weight_key: str,
+        dataset_reader: DatasetReader,
+        data_path: str,
+        model: Lazy[Model],
+        data_loader: Lazy[DataLoader],
+        trainer: Lazy[Trainer],
+        cross_validator: CrossValidator,
+        local_rank: int,  # It's passed transparently directly to the trainer.
+        vocabulary: Lazy[Vocabulary] = None,
+        retrain: bool = False,
+    ) -> "CrossValidateModel":
+        logger.info(f"Reading data from {data_path}")
+        dataset = dataset_reader.read(data_path)
+
+        vocabulary_ = vocabulary.construct(instances=dataset)
+        if not vocabulary_:
+            vocabulary_ = Vocabulary.from_instances(dataset)
+        model_ = model.construct(vocab=vocabulary_)
+
+        if common_util.is_master():
```

Comment:
Pretty sure you can remove these, unless you are actually planning on literally using the `train` script for this, including the distributed training.  Not sure how that interacts with how this does cross validation.

Reply:
> unless you are actually planning on literally using the `train` script for this,

You mean the parts of `train.py` that are not in `TrainModel`?

Reply:
Yes, that's what I mean.

Reply:
I think it *could* work in distributed training as it is with some changes I just added. Though I haven't tried it. So I'll leave it.